iopath/common/gcs.py

```diff
@@ -0,0 +1,281 @@
+        Args:
+            path (str): A URI supported by this PathHandler
+        Returns:
+            List[str]: list of contents in given path
+        """
+        self._check_kwargs(kwargs)
+        bucket_name, gcs_path = self._parse_uri(path)
+        client = self._get_client()
+        return [
+            f"{blob.name}"
+            for blob in client.list_blobs(bucket_name, prefix=gcs_path, delimiter="/")
+            if blob.name != gcs_path  # exclude dirname
+        ]
+
+    def _rm(self, path: str, **kwargs: Any) -> None:
+        """
+        Remove the file (not directory) at the provided URI.
+        Args:
+            path (str): A URI supported by this PathHandler
+        """
+        self._check_kwargs(kwargs)
+
+        blob = self._get_blob(path)
+        blob.delete()
+
+    def _mkdirs(self, path: str, **kwargs: Any) -> None:
+        """
+        Recursive directory creation function. Like mkdir(), but makes all
+        intermediate-level directories needed to contain the leaf directory.
+        Similar to the native `os.makedirs`.
+        Args:
+            path (str): A URI supported by this PathHandler
+        """
+        self._check_kwargs(kwargs)
+
+        assert path.endswith("/"), path
+
+        destination_blob = self._get_blob(path)
+        destination_blob.upload_from_string(
+            ""
+        )  # https://stackoverflow.com/questions/38416598/how-to-create-an-empty-folder-on-google-storage-with-google-api
+
+    def _open(
+        self,
+        path: str,
+        mode: str = "r",
+        # The following three arguments are unused,
+        # But are included to avoid triggering WARNING
+        # messages from _check_kargs.
+        buffering: int = -1,
+        encoding: Optional[str] = None,
+        errors: Optional[str] = None,
+        newline: Optional[str] = None,
+        read_chunk_size: Optional[int] = None,
+        **kwargs: Any,
+    ) -> Union[IO[str], IO[bytes]]:
+        """
+        Open a stream to a URI, similar to the built-in `open`.
+        Args:
+            path (str): A URI supported by this PathHandler
+            mode (str): Specifies the mode in which the file is opened. It defaults
+                to 'r'.
+            buffering (int): An optional integer used to set the buffering policy.
+                Pass 0 to switch buffering off and an integer >= 1 to indicate the
+                size in bytes of a fixed-size chunk buffer. When no buffering
+                argument is given, the default buffering policy depends on the
+                underlying I/O implementation.
+        Returns:
+            file: a file-like object.
+        """
+        self._check_kwargs(kwargs)
+        assert (
+            buffering == -1
+        ), f"{self.__class__.__name__} does not support the `buffering` argument"
+
+        if "r" in mode:
+            local_path = self._get_local_path(path)
+            return open(local_path, mode)
```

Comment:
It should use `blob.open` instead of caching. Caching has the side effect that it may fill up the disk space, so it's not a good idea to transparently apply caching when reading. Instead, users should explicit call for caching if needed, by using `get_local_path`.

We apply this strategy to some internal storage backends as well.